src/python/TaskWorker/Actions/DBSDataDiscovery.py

```diff
@@ -572,6 +403,232 @@ def executeInternal(self, *args, **kwargs):
+        msg = msgHead
+        if not isinstance(dataToRecall, str) and not isinstance(dataToRecall, list):
+            return
+        if system == 'Rucio':
+            self.requestRucioTapeRecall(dataToRecall, sizeToRecall, tapeLocations, msgHead=msg)
+            msg += "\nThis task will be automatically submitted as soon as the stage-out is completed."
+            self.uploadWarning(msg, self.userproxy, self.taskName)
+            raise TapeDatasetException(msg)
+        elif system == 'None':
+            msg += '\nIt is not possible to request a recall from tape.'
+            msg += "\nPlease, check DAS (https://cmsweb.cern.ch/das) and make sure the dataset is accessible on DISK."
+            raise TaskWorkerException(msg)
+
+    def requestRucioTapeRecall(self, dataToRecall=None, sizeToRecall=0, tapeLocations=None, msgHead=''):
+        """
+        implements tape recall with Rucio
+        params and returns are the same as requestTapeRecall defined above
+        """
+        # args to replace self: config, taskName, logger, crabserver
+        msg = msgHead
+        # need to use crab_tape_recall Rucio account to create containers and create rules
+        tapeRecallConfig = copy.copy(self.config)
+        tapeRecallConfig.Services.Rucio_account = 'crab_tape_recall'
+        rucioClient = getNativeRucioClient(tapeRecallConfig, self.logger)  # pylint: disable=redefined-outer-name
+
+        # prepare container to be recalled
+        if isinstance(dataToRecall, str):
+            # recalling a full DBS dataset, simple the container already exists
+            myScope = 'cms'
+            dbsDatasetName = dataToRecall
+            containerDid = {'scope': myScope, 'name': dbsDatasetName}
+        elif isinstance(dataToRecall, list):
+            # need to prepare ad hoc container. Can not reuse a possibly existing one:
+            # if a user wants one block from a large dataset, we must
+            # not recall also all blocks possibly requested in the past
+            #  by other users for other reasons.
+            #  Therefore we stick with naming the container after the task name
+            myScope = 'user.crab_tape_recall'  # do not mess up with cms scope
+            containerName = '/TapeRecall/%s/USER' % self.taskName.replace(':', '.')
+            containerDid = {'scope': myScope, 'name': containerName}
+            self.makeContainerFromBlockList(
+                rucio=rucioClient, blockList=dataToRecall,
+                containerDid=containerDid, logger=self.logger)
+            # beware blockList being not subscriptable (e.g. dict_keys type)
+            dbsDatasetName = next(iter(dataToRecall)).split('#')[0]
+        else:
+            return
+
+        TBtoRecall = sizeToRecall // 1e12
+        # a friendly-formatted string to print the size
+        recallSize = f"{TBtoRecall:.0f} TBytes" if TBtoRecall > 0 else f"{sizeToRecall / 1e9:.0f} GBytes"
+        # Sanity check
+        if TBtoRecall > 1e3:
```

Comment:
no, I am not planning to recall data larger than 1PB. The limit at 1PB is mostly a safety in case something goes wrong elsewhere.
But indeed using it in two places like this is not good. 
The value in Line 557 may need to be changed instead. Hopefully not, but it may. I am hesitant to put it in the configuration file. Let's add MAX_TB_TO_RECALL_AT_A_SINGLE_SITE to ServerUtilities

Reply:
Ah, sorry. I did not open the link in the comment on line 557. I agree to put  `MAX_TB_TO_RECALL_AT_A_SINGLE_SITE` to ServerUtils.